In [1]:

import pandas as pd
import numpy as np
import re
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import scipy.stats as stats


In [2]:
df_final_demo = r'https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_demo.txt'
df_final_web_data_pt1 = r'https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_web_data_pt_1.txt'
df_final_web_data_pt2 = r'https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_web_data_pt_2.txt'
df_final_experiment_client = r'https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_experiment_clients.txt'



In [3]:
# Intenta leer el archivo
try:
    df_final_demo = pd.read_csv(df_final_demo, sep=',')  
    df_final_web_data_pt1= pd.read_csv(df_final_web_data_pt1, sep=',')
    df_final_web_data_pt2= pd.read_csv(df_final_web_data_pt2, sep=',')
    df_final_experiment_client= pd.read_csv(df_final_experiment_client, sep=',')
except Exception as e:
    print(f"Error al cargar el archivo: {e}")

In [4]:
df_final_demo

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0         836976             6.0              73.0      60.5     U        2.0   
1        2304905             7.0              94.0      58.0     U        2.0   
2        1439522             5.0              64.0      32.0     U        2.0   
3        1562045            16.0             198.0      49.0     M        2.0   
4        5126305            12.0             145.0      33.0     F        2.0   
...          ...             ...               ...       ...   ...        ...   
70604    7993686             4.0              56.0      38.5     U        3.0   
70605    8981690            12.0             148.0      31.0     M        2.0   
70606     333913            16.0             198.0      61.5     F        2.0   
70607    1573142            21.0             255.0      68.0     M        3.0   
70608    5602139            21.0             254.0      59.5     F        3.0   

              bal  calls_6_mnth  logons_6_mnth  
0        45105.30           6.0            9.0  
1       110860.30           6.0            9.0  
2        52467.79           6.0            9.0  
3        67454.65           3.0            6.0  
4       103671.75           0.0            3.0  
...           ...           ...            ...  
70604  1411062.68           5.0            5.0  
70605   101867.07           6.0            6.0  
70606    40745.00           3.0            3.0  
70607   475114.69           4.0            4.0  
70608   157498.73           7.0            7.0  

[70609 rows x 9 columns]

In [5]:
df_final_demo.columns = df_final_demo.columns.str.lower()
df_final_web_data_pt1.columns = df_final_web_data_pt1.columns.str.lower()
df_final_web_data_pt2.columns = df_final_web_data_pt2.columns.str.lower()
df_final_experiment_client.columns = df_final_experiment_client.columns.str.lower()


In [6]:
final_demo = df_final_demo
final_demo

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0         836976             6.0              73.0      60.5     U        2.0   
1        2304905             7.0              94.0      58.0     U        2.0   
2        1439522             5.0              64.0      32.0     U        2.0   
3        1562045            16.0             198.0      49.0     M        2.0   
4        5126305            12.0             145.0      33.0     F        2.0   
...          ...             ...               ...       ...   ...        ...   
70604    7993686             4.0              56.0      38.5     U        3.0   
70605    8981690            12.0             148.0      31.0     M        2.0   
70606     333913            16.0             198.0      61.5     F        2.0   
70607    1573142            21.0             255.0      68.0     M        3.0   
70608    5602139            21.0             254.0      59.5     F        3.0   

              bal  calls_6_mnth  logons_6_mnth  
0        45105.30           6.0            9.0  
1       110860.30           6.0            9.0  
2        52467.79           6.0            9.0  
3        67454.65           3.0            6.0  
4       103671.75           0.0            3.0  
...           ...           ...            ...  
70604  1411062.68           5.0            5.0  
70605   101867.07           6.0            6.0  
70606    40745.00           3.0            3.0  
70607   475114.69           4.0            4.0  
70608   157498.73           7.0            7.0  

[70609 rows x 9 columns]

In [7]:
final_demo = pd.merge(final_demo, df_final_experiment_client, on='client_id')
final_demo

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0         836976             6.0              73.0      60.5     U        2.0   
1        2304905             7.0              94.0      58.0     U        2.0   
2        1439522             5.0              64.0      32.0     U        2.0   
3        1562045            16.0             198.0      49.0     M        2.0   
4        5126305            12.0             145.0      33.0     F        2.0   
...          ...             ...               ...       ...   ...        ...   
70604    7993686             4.0              56.0      38.5     U        3.0   
70605    8981690            12.0             148.0      31.0     M        2.0   
70606     333913            16.0             198.0      61.5     F        2.0   
70607    1573142            21.0             255.0      68.0     M        3.0   
70608    5602139            21.0             254.0      59.5     F        3.0   

              bal  calls_6_mnth  logons_6_mnth variation  
0        45105.30           6.0            9.0      Test  
1       110860.30           6.0            9.0   Control  
2        52467.79           6.0            9.0      Test  
3        67454.65           3.0            6.0      Test  
4       103671.75           0.0            3.0   Control  
...           ...           ...            ...       ...  
70604  1411062.68           5.0            5.0       NaN  
70605   101867.07           6.0            6.0       NaN  
70606    40745.00           3.0            3.0       NaN  
70607   475114.69           4.0            4.0       NaN  
70608   157498.73           7.0            7.0       NaN  

[70609 rows x 10 columns]

In [8]:
final_demo = final_demo.dropna()
final_demo

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0         836976             6.0              73.0      60.5     U        2.0   
1        2304905             7.0              94.0      58.0     U        2.0   
2        1439522             5.0              64.0      32.0     U        2.0   
3        1562045            16.0             198.0      49.0     M        2.0   
4        5126305            12.0             145.0      33.0     F        2.0   
...          ...             ...               ...       ...   ...        ...   
50495    1780858            21.0             262.0      68.5     M        3.0   
50496    6967120            21.0             260.0      68.5     M        3.0   
50497    5826160            20.0             249.0      56.5     F        2.0   
50498    8739285            19.0             229.0      69.5     F        2.0   
50499    7468138            18.0             222.0      61.0     F        3.0   

              bal  calls_6_mnth  logons_6_mnth variation  
0        45105.30           6.0            9.0      Test  
1       110860.30           6.0            9.0   Control  
2        52467.79           6.0            9.0      Test  
3        67454.65           3.0            6.0      Test  
4       103671.75           0.0            3.0   Control  
...           ...           ...            ...       ...  
50495   372100.59           6.0            9.0      Test  
50496  4279873.38           6.0            9.0   Control  
50497    44837.16           2.0            5.0      Test  
50498    44994.24           1.0            4.0      Test  
50499   209278.15           0.0            3.0      Test  

[50487 rows x 10 columns]

In [9]:
df_final_web_data_pt1

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
343136    2443347  465784886_73090545671  136329900_10529659391_316129   
343137    2443347  465784886_73090545671  136329900_10529659391_316129   
343138    2443347  465784886_73090545671  136329900_10529659391_316129   
343139    2443347  465784886_73090545671  136329900_10529659391_316129   
343140    2443347  465784886_73090545671  136329900_10529659391_316129   

       process_step            date_time  
0            step_3  2017-04-17 15:27:07  
1            step_2  2017-04-17 15:26:51  
2            step_3  2017-04-17 15:19:22  
3            step_2  2017-04-17 15:19:13  
4            step_3  2017-04-17 15:18:04  
...             ...                  ...  
343136      confirm  2017-03-31 15:15:46  
343137       step_3  2017-03-31 15:14:53  
343138       step_2  2017-03-31 15:12:08  
343139       step_1  2017-03-31 15:11:37  
343140        start  2017-03-31 15:10:47  

[343141 rows x 5 columns]

In [10]:
df_final_web_data_pt2

client_id             visitor_id                      visit_id  \
0          763412  601952081_10457207388  397475557_40440946728_419634   
1         6019349  442094451_91531546617  154620534_35331068705_522317   
2         6019349  442094451_91531546617  154620534_35331068705_522317   
3         6019349  442094451_91531546617  154620534_35331068705_522317   
4         6019349  442094451_91531546617  154620534_35331068705_522317   
...           ...                    ...                           ...   
412259    9668240   388766751_9038881013   922267647_3096648104_968866   
412260    9668240   388766751_9038881013   922267647_3096648104_968866   
412261    9668240   388766751_9038881013   922267647_3096648104_968866   
412262    9668240   388766751_9038881013   922267647_3096648104_968866   
412263     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step            date_time  
0           confirm  2017-06-06 08:56:00  
1           confirm  2017-06-01 11:59:27  
2            step_3  2017-06-01 11:58:48  
3            step_2  2017-06-01 11:58:08  
4            step_1  2017-06-01 11:57:58  
...             ...                  ...  
412259        start  2017-05-24 18:46:10  
412260        start  2017-05-24 18:45:29  
412261       step_1  2017-05-24 18:44:51  
412262        start  2017-05-24 18:44:34  
412263        start  2017-06-03 12:17:09  

[412264 rows x 5 columns]

In [11]:
final_web = pd.concat([df_final_web_data_pt1, df_final_web_data_pt2], ignore_index=True)
final_web

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
755400    9668240   388766751_9038881013   922267647_3096648104_968866   
755401    9668240   388766751_9038881013   922267647_3096648104_968866   
755402    9668240   388766751_9038881013   922267647_3096648104_968866   
755403    9668240   388766751_9038881013   922267647_3096648104_968866   
755404     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step            date_time  
0            step_3  2017-04-17 15:27:07  
1            step_2  2017-04-17 15:26:51  
2            step_3  2017-04-17 15:19:22  
3            step_2  2017-04-17 15:19:13  
4            step_3  2017-04-17 15:18:04  
...             ...                  ...  
755400        start  2017-05-24 18:46:10  
755401        start  2017-05-24 18:45:29  
755402       step_1  2017-05-24 18:44:51  
755403        start  2017-05-24 18:44:34  
755404        start  2017-06-03 12:17:09  

[755405 rows x 5 columns]

In [12]:
final_web.isna()

client_id  visitor_id  visit_id  process_step  date_time
0           False       False     False         False      False
1           False       False     False         False      False
2           False       False     False         False      False
3           False       False     False         False      False
4           False       False     False         False      False
...           ...         ...       ...           ...        ...
755400      False       False     False         False      False
755401      False       False     False         False      False
755402      False       False     False         False      False
755403      False       False     False         False      False
755404      False       False     False         False      False

[755405 rows x 5 columns]

In [13]:
print(final_web.isnull().sum())


client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64


In [14]:
tabla_analisis = pd.merge(final_demo, final_web, on='client_id')

In [15]:
tabla_analisis

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  \
0          836976             6.0              73.0      60.5     U   
1          836976             6.0              73.0      60.5     U   
2          836976             6.0              73.0      60.5     U   
3          836976             6.0              73.0      60.5     U   
4          836976             6.0              73.0      60.5     U   
...           ...             ...               ...       ...   ...   
321190    7468138            18.0             222.0      61.0     F   
321191    7468138            18.0             222.0      61.0     F   
321192    7468138            18.0             222.0      61.0     F   
321193    7468138            18.0             222.0      61.0     F   
321194    7468138            18.0             222.0      61.0     F   

        num_accts        bal  calls_6_mnth  logons_6_mnth variation  \
0             2.0   45105.30           6.0            9.0      Test   
1             2.0   45105.30           6.0            9.0      Test   
2             2.0   45105.30           6.0            9.0      Test   
3             2.0   45105.30           6.0            9.0      Test   
4             2.0   45105.30           6.0            9.0      Test   
...           ...        ...           ...            ...       ...   
321190        3.0  209278.15           0.0            3.0      Test   
321191        3.0  209278.15           0.0            3.0      Test   
321192        3.0  209278.15           0.0            3.0      Test   
321193        3.0  209278.15           0.0            3.0      Test   
321194        3.0  209278.15           0.0            3.0      Test   

                   visitor_id                      visit_id process_step  \
0        427070339_1413275162   228976764_46825473280_96584      confirm   
1        427070339_1413275162   228976764_46825473280_96584      confirm   
2        427070339_1413275162   228976764_46825473280_96584      confirm   
3        427070339_1413275162   228976764_46825473280_96584       step_3   
4        427070339_1413275162   228976764_46825473280_96584       step_2   
...                       ...                           ...          ...   
321190  317123960_46820919455  769876461_30381166055_830233       step_2   
321191  317123960_46820919455  769876461_30381166055_830233       step_1   
321192  317123960_46820919455  769876461_30381166055_830233        start   
321193  317123960_46820919455  769876461_30381166055_830233        start   
321194  317123960_46820919455  769876461_30381166055_830233        start   

                  date_time  
0       2017-04-02 11:51:13  
1       2017-04-02 11:47:50  
2       2017-04-02 11:46:45  
3       2017-04-02 11:23:08  
4       2017-04-02 11:22:24  
...                     ...  
321190  2017-03-30 23:59:15  
321191  2017-03-30 23:58:51  
321192  2017-03-30 23:58:40  
321193  2017-03-30 23:55:11  
321194  2017-03-31 00:16:12  

[321195 rows x 14 columns]

In [16]:
tabla_analisis.head()

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0     836976             6.0              73.0      60.5     U        2.0   
1     836976             6.0              73.0      60.5     U        2.0   
2     836976             6.0              73.0      60.5     U        2.0   
3     836976             6.0              73.0      60.5     U        2.0   
4     836976             6.0              73.0      60.5     U        2.0   

       bal  calls_6_mnth  logons_6_mnth variation            visitor_id  \
0  45105.3           6.0            9.0      Test  427070339_1413275162   
1  45105.3           6.0            9.0      Test  427070339_1413275162   
2  45105.3           6.0            9.0      Test  427070339_1413275162   
3  45105.3           6.0            9.0      Test  427070339_1413275162   
4  45105.3           6.0            9.0      Test  427070339_1413275162   

                      visit_id process_step            date_time  
0  228976764_46825473280_96584      confirm  2017-04-02 11:51:13  
1  228976764_46825473280_96584      confirm  2017-04-02 11:47:50  
2  228976764_46825473280_96584      confirm  2017-04-02 11:46:45  
3  228976764_46825473280_96584       step_3  2017-04-02 11:23:08  
4  228976764_46825473280_96584       step_2  2017-04-02 11:22:24

In [17]:
tabla_analisis["process_step"].unique()

array(['confirm', 'step_3', 'step_2', 'step_1', 'start'], dtype=object)

In [18]:
tabla_test = tabla_analisis.set_index(['client_id', 'process_step'])
tabla_test.tail(10)

clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  \
client_id process_step                                                     
8739285   step_3                  19.0             229.0      69.5     F   
          step_2                  19.0             229.0      69.5     F   
          step_1                  19.0             229.0      69.5     F   
          start                   19.0             229.0      69.5     F   
7468138   step_3                  18.0             222.0      61.0     F   
          step_2                  18.0             222.0      61.0     F   
          step_1                  18.0             222.0      61.0     F   
          start                   18.0             222.0      61.0     F   
          start                   18.0             222.0      61.0     F   
          start                   18.0             222.0      61.0     F   

                        num_accts        bal  calls_6_mnth  logons_6_mnth  \
client_id process_step                                                      
8739285   step_3              2.0   44994.24           1.0            4.0   
          step_2              2.0   44994.24           1.0            4.0   
          step_1              2.0   44994.24           1.0            4.0   
          start               2.0   44994.24           1.0            4.0   
7468138   step_3              3.0  209278.15           0.0            3.0   
          step_2              3.0  209278.15           0.0            3.0   
          step_1              3.0  209278.15           0.0            3.0   
          start               3.0  209278.15           0.0            3.0   
          start               3.0  209278.15           0.0            3.0   
          start               3.0  209278.15           0.0            3.0   

                       variation             visitor_id  \
client_id process_step                                    
8739285   step_3            Test    8756552_54841419737   
          step_2            Test    8756552_54841419737   
          step_1            Test    8756552_54841419737   
          start             Test    8756552_54841419737   
7468138   step_3            Test  317123960_46820919455   
          step_2            Test  317123960_46820919455   
          step_1            Test  317123960_46820919455   
          start             Test  317123960_46820919455   
          start             Test  317123960_46820919455   
          start             Test  317123960_46820919455   

                                            visit_id            date_time  
client_id process_step                                                     
8739285   step_3        101846948_96925910026_299041  2017-04-12 14:58:39  
          step_2        101846948_96925910026_299041  2017-04-12 14:58:16  
          step_1        101846948_96925910026_299041  2017-04-12 14:50:44  
          start         101846948_96925910026_299041  2017-04-12 14:49:47  
7468138   step_3        769876461_30381166055_830233  2017-03-30 23:59:33  
          step_2        769876461_30381166055_830233  2017-03-30 23:59:15  
          step_1        769876461_30381166055_830233  2017-03-30 23:58:51  
          start         769876461_30381166055_830233  2017-03-30 23:58:40  
          start         769876461_30381166055_830233  2017-03-30 23:55:11  
          start         769876461_30381166055_830233  2017-03-31 00:16:12

In [19]:
tabla_test

clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  \
client_id process_step                                                     
836976    confirm                  6.0              73.0      60.5     U   
          confirm                  6.0              73.0      60.5     U   
          confirm                  6.0              73.0      60.5     U   
          step_3                   6.0              73.0      60.5     U   
          step_2                   6.0              73.0      60.5     U   
...                                ...               ...       ...   ...   
7468138   step_2                  18.0             222.0      61.0     F   
          step_1                  18.0             222.0      61.0     F   
          start                   18.0             222.0      61.0     F   
          start                   18.0             222.0      61.0     F   
          start                   18.0             222.0      61.0     F   

                        num_accts        bal  calls_6_mnth  logons_6_mnth  \
client_id process_step                                                      
836976    confirm             2.0   45105.30           6.0            9.0   
          confirm             2.0   45105.30           6.0            9.0   
          confirm             2.0   45105.30           6.0            9.0   
          step_3              2.0   45105.30           6.0            9.0   
          step_2              2.0   45105.30           6.0            9.0   
...                           ...        ...           ...            ...   
7468138   step_2              3.0  209278.15           0.0            3.0   
          step_1              3.0  209278.15           0.0            3.0   
          start               3.0  209278.15           0.0            3.0   
          start               3.0  209278.15           0.0            3.0   
          start               3.0  209278.15           0.0            3.0   

                       variation             visitor_id  \
client_id process_step                                    
836976    confirm           Test   427070339_1413275162   
          confirm           Test   427070339_1413275162   
          confirm           Test   427070339_1413275162   
          step_3            Test   427070339_1413275162   
          step_2            Test   427070339_1413275162   
...                          ...                    ...   
7468138   step_2            Test  317123960_46820919455   
          step_1            Test  317123960_46820919455   
          start             Test  317123960_46820919455   
          start             Test  317123960_46820919455   
          start             Test  317123960_46820919455   

                                            visit_id            date_time  
client_id process_step                                                     
836976    confirm        228976764_46825473280_96584  2017-04-02 11:51:13  
          confirm        228976764_46825473280_96584  2017-04-02 11:47:50  
          confirm        228976764_46825473280_96584  2017-04-02 11:46:45  
          step_3         228976764_46825473280_96584  2017-04-02 11:23:08  
          step_2         228976764_46825473280_96584  2017-04-02 11:22:24  
...                                              ...                  ...  
7468138   step_2        769876461_30381166055_830233  2017-03-30 23:59:15  
          step_1        769876461_30381166055_830233  2017-03-30 23:58:51  
          start         769876461_30381166055_830233  2017-03-30 23:58:40  
          start         769876461_30381166055_830233  2017-03-30 23:55:11  
          start         769876461_30381166055_830233  2017-03-31 00:16:12  

[321195 rows x 12 columns]

In [20]:
tabla_analisis_reducida = tabla_analisis[['client_id', 'process_step']]
tabla_analisis_reducida

client_id process_step
0          836976      confirm
1          836976      confirm
2          836976      confirm
3          836976       step_3
4          836976       step_2
...           ...          ...
321190    7468138       step_2
321191    7468138       step_1
321192    7468138        start
321193    7468138        start
321194    7468138        start

[321195 rows x 2 columns]

In [21]:
pivot_df = tabla_analisis.pivot_table(index=['client_id', 'visitor_id', 'visit_id'], columns=['process_step'], aggfunc='size', fill_value=0)
pivot_df

process_step                                                  confirm  start  \
client_id visitor_id            visit_id                                       
555       402506806_56087378777 637149525_38041617439_716659        1      1   
647       66758770_53988066587  40369564_40101682850_311847         1      1   
934       810392784_45004760546 7076463_57954418406_971348          0      4   
1028      42237450_62128060588  557292053_87239438319_391157        0      1   
1104      194240915_18158000533 543158812_46395476577_767725        0      1   
...                                                               ...    ...   
9999626   52633065_71189986073  182723364_27031318473_880901        0      1   
9999729   604429154_69247391147 99583652_41711450505_426179         0      1   
          834634258_21862004160 870243567_56915814033_814203        1      1   
          843385170_36953471821 493310979_9209676464_421146         0      2   
9999832   145538019_54444341400 472154369_16714624241_585315        0      1   

process_step                                                  step_1  step_2  \
client_id visitor_id            visit_id                                       
555       402506806_56087378777 637149525_38041617439_716659       1       1   
647       66758770_53988066587  40369564_40101682850_311847        1       1   
934       810392784_45004760546 7076463_57954418406_971348         0       0   
1028      42237450_62128060588  557292053_87239438319_391157       5       2   
1104      194240915_18158000533 543158812_46395476577_767725       0       0   
...                                                              ...     ...   
9999626   52633065_71189986073  182723364_27031318473_880901       1       0   
9999729   604429154_69247391147 99583652_41711450505_426179        1       0   
          834634258_21862004160 870243567_56915814033_814203       1       1   
          843385170_36953471821 493310979_9209676464_421146        1       1   
9999832   145538019_54444341400 472154369_16714624241_585315       1       0   

process_step                                                  step_3  
client_id visitor_id            visit_id                              
555       402506806_56087378777 637149525_38041617439_716659       1  
647       66758770_53988066587  40369564_40101682850_311847        1  
934       810392784_45004760546 7076463_57954418406_971348         0  
1028      42237450_62128060588  557292053_87239438319_391157       1  
1104      194240915_18158000533 543158812_46395476577_767725       0  
...                                                              ...  
9999626   52633065_71189986073  182723364_27031318473_880901       0  
9999729   604429154_69247391147 99583652_41711450505_426179        0  
          834634258_21862004160 870243567_56915814033_814203       1  
          843385170_36953471821 493310979_9209676464_421146        0  
9999832   145538019_54444341400 472154369_16714624241_585315       0  

[69425 rows x 5 columns]

In [22]:
# Get the number of unique client_id values in the pivot_df
unique_client_ids = pivot_df.index.get_level_values('client_id').nunique()

# Display the number of unique client_id values
print(f"Number of unique client_id values: {unique_client_ids}")

Number of unique client_id values: 50487


In [23]:
# Get the number of unique client_id values in the pivot_df
unique_client_visitors_ids = pivot_df.index.get_level_values('visitor_id').nunique()

# Display the number of unique client_id values
print(f"Number of unique visitor_id values: {unique_client_visitors_ids}")

Number of unique visitor_id values: 55994


In [24]:
# Step 2: Group by 'client_id' and count unique 'visit_id' values
visit_counts = tabla_analisis.groupby('client_id')['visit_id'].nunique().reset_index()
visit_counts.columns = ['client_id', 'visit_count']
visit_counts

# Step 3: Merge the unique visit counts back to the original DataFrame
pivot_df = pivot_df.merge(visit_counts, on='client_id', how='left')

# Display the DataFrame with the new column
pivot_df

client_id  confirm  start  step_1  step_2  step_3  visit_count
0            555        1      1       1       1       1            1
1            647        1      1       1       1       1            1
2            934        0      4       0       0       0            1
3           1028        0      1       5       2       1            1
4           1104        0      1       0       0       0            2
...          ...      ...    ...     ...     ...     ...          ...
69420    9999626        0      1       1       0       0            1
69421    9999729        0      1       1       0       0            3
69422    9999729        1      1       1       1       1            3
69423    9999729        0      2       1       1       0            3
69424    9999832        0      1       1       0       0            1

[69425 rows x 7 columns]

In [25]:
# Ensure 'client_id' is in both DataFrames
if 'client_id' not in pivot_df.columns:
    pivot_df = pivot_df.reset_index()

# Select the necessary columns from tabla_analisis
variation_df = tabla_analisis[['client_id', 'variation']].drop_duplicates()

# Merge the variation column to the pivot_df
pivot_df = pivot_df.merge(variation_df, on='client_id', how='left')

# Display the updated pivot_df
pivot_df

client_id  confirm  start  step_1  step_2  step_3  visit_count  \
0            555        1      1       1       1       1            1   
1            647        1      1       1       1       1            1   
2            934        0      4       0       0       0            1   
3           1028        0      1       5       2       1            1   
4           1104        0      1       0       0       0            2   
...          ...      ...    ...     ...     ...     ...          ...   
69420    9999626        0      1       1       0       0            1   
69421    9999729        0      1       1       0       0            3   
69422    9999729        1      1       1       1       1            3   
69423    9999729        0      2       1       1       0            3   
69424    9999832        0      1       1       0       0            1   

      variation  
0          Test  
1          Test  
2          Test  
3       Control  
4       Control  
...         ...  
69420      Test  
69421      Test  
69422      Test  
69423      Test  
69424      Test  

[69425 rows x 8 columns]

In [26]:
# Define the new column order
new_column_order = ['client_id', 'start', 'step_1', 'step_2', 'step_3', 'confirm', 'visit_count', 'variation']

# Reorder the columns
pivot_df = pivot_df[new_column_order]

# Display the updated DataFrame
pivot_df

client_id  start  step_1  step_2  step_3  confirm  visit_count  \
0            555      1       1       1       1        1            1   
1            647      1       1       1       1        1            1   
2            934      4       0       0       0        0            1   
3           1028      1       5       2       1        0            1   
4           1104      1       0       0       0        0            2   
...          ...    ...     ...     ...     ...      ...          ...   
69420    9999626      1       1       0       0        0            1   
69421    9999729      1       1       0       0        0            3   
69422    9999729      1       1       1       1        1            3   
69423    9999729      2       1       1       0        0            3   
69424    9999832      1       1       0       0        0            1   

      variation  
0          Test  
1          Test  
2          Test  
3       Control  
4       Control  
...         ...  
69420      Test  
69421      Test  
69422      Test  
69423      Test  
69424      Test  

[69425 rows x 8 columns]

In [27]:
# Step 1: Convert the 'date_time' column to datetime
tabla_analisis['date_time'] = pd.to_datetime(tabla_analisis['date_time'])

# Step 2: Sort the DataFrame by 'visit_id' and 'date_time'
tabla_analisis = tabla_analisis.sort_values(by=['visit_id', 'date_time'])

# Step 3: Calculate the time between steps
tabla_analisis['time_in_step'] = (
    tabla_analisis.groupby('visit_id')['date_time'].shift(-1) - tabla_analisis['date_time']
).dt.total_seconds()

# Step 4: Calculate the total time spent in each step
time_in_steps = tabla_analisis.groupby(['visit_id', 'process_step'])['time_in_step'].sum().reset_index()

# Step 5: Pivot the DataFrame to get the total time spent in each step for each visit_id
time_pivot = time_in_steps.pivot(index='visit_id', columns='process_step', values='time_in_step').fillna(0)

# Step 6: Merge the pivot table back to the original DataFrame if needed
tabla_analisis = tabla_analisis.merge(time_pivot, on='visit_id', how='left', suffixes=('', '_total_time'))

# Display the updated DataFrame
tabla_analisis

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  \
0         3561384             4.0              56.0      59.5     U   
1         3561384             4.0              56.0      59.5     U   
2         7338123             7.0              88.0      23.5     M   
3         7338123             7.0              88.0      23.5     M   
4         7338123             7.0              88.0      23.5     M   
...           ...             ...               ...       ...   ...   
321190    6627522            18.0             227.0      58.0     F   
321191    6627522            18.0             227.0      58.0     F   
321192    6627522            18.0             227.0      58.0     F   
321193    6627522            18.0             227.0      58.0     F   
321194    6627522            18.0             227.0      58.0     F   

        num_accts         bal  calls_6_mnth  logons_6_mnth variation  \
0             2.0    63130.44           6.0            9.0      Test   
1             2.0    63130.44           6.0            9.0      Test   
2             2.0    26436.73           6.0            9.0      Test   
3             2.0    26436.73           6.0            9.0      Test   
4             2.0    26436.73           6.0            9.0      Test   
...           ...         ...           ...            ...       ...   
321190        2.0  1056775.54           6.0            9.0      Test   
321191        2.0  1056775.54           6.0            9.0      Test   
321192        2.0  1056775.54           6.0            9.0      Test   
321193        2.0  1056775.54           6.0            9.0      Test   
321194        2.0  1056775.54           6.0            9.0      Test   

                   visitor_id                      visit_id process_step  \
0        451664975_1722933822  100012776_37918976071_457913      confirm   
1        451664975_1722933822  100012776_37918976071_457913      confirm   
2       612065484_94198474375   100019538_17884295066_43909        start   
3       612065484_94198474375   100019538_17884295066_43909       step_1   
4       612065484_94198474375   100019538_17884295066_43909       step_2   
...                       ...                           ...          ...   
321190  730634087_44272418812  999988789_76411676596_272843        start   
321191  730634087_44272418812  999988789_76411676596_272843       step_1   
321192  730634087_44272418812  999988789_76411676596_272843       step_2   
321193  730634087_44272418812  999988789_76411676596_272843       step_1   
321194  730634087_44272418812  999988789_76411676596_272843        start   

                 date_time  time_in_step  confirm  start  step_1  step_2  \
0      2017-04-26 13:22:17          52.0     52.0    0.0     0.0     0.0   
1      2017-04-26 13:23:09           NaN     52.0    0.0     0.0     0.0   
2      2017-04-09 16:20:56          16.0      0.0   35.0    23.0   127.0   
3      2017-04-09 16:21:12           9.0      0.0   35.0    23.0   127.0   
4      2017-04-09 16:21:21          14.0      0.0   35.0    23.0   127.0   
...                    ...           ...      ...    ...     ...     ...   
321190 2017-04-21 23:49:11          11.0      0.0   11.0    63.0    44.0   
321191 2017-04-21 23:49:22          54.0      0.0   11.0    63.0    44.0   
321192 2017-04-21 23:50:16          44.0      0.0   11.0    63.0    44.0   
321193 2017-04-21 23:51:00           9.0      0.0   11.0    63.0    44.0   
321194 2017-04-21 23:51:09           NaN      0.0   11.0    63.0    44.0   

        step_3  
0          0.0  
1          0.0  
2         57.0  
3         57.0  
4         57.0  
...        ...  
321190     0.0  
321191     0.0  
321192     0.0  
321193     0.0  
321194     0.0  

[321195 rows x 20 columns]

In [28]:
pivot_df.rename(columns={
    'start': 'start_counts',
    'step_1': 'step_1_counts',
    'step_2': 'step_2_counts',
    'step_3': 'step_3_counts',
    'confirm': 'confirm_counts'
}, inplace=True)

pivot_df


client_id  start_counts  step_1_counts  step_2_counts  step_3_counts  \
0            555             1              1              1              1   
1            647             1              1              1              1   
2            934             4              0              0              0   
3           1028             1              5              2              1   
4           1104             1              0              0              0   
...          ...           ...            ...            ...            ...   
69420    9999626             1              1              0              0   
69421    9999729             1              1              0              0   
69422    9999729             1              1              1              1   
69423    9999729             2              1              1              0   
69424    9999832             1              1              0              0   

       confirm_counts  visit_count variation  
0                   1            1      Test  
1                   1            1      Test  
2                   0            1      Test  
3                   0            1   Control  
4                   0            2   Control  
...               ...          ...       ...  
69420               0            1      Test  
69421               0            3      Test  
69422               1            3      Test  
69423               0            3      Test  
69424               0            1      Test  

[69425 rows x 8 columns]

In [29]:
# Ensure 'client_id' is in both DataFrames
if 'client_id' not in pivot_df.columns:
    pivot_df = pivot_df.reset_index()

# Select the necessary columns from tabla_analisis
step_time_df = tabla_analisis[['client_id','start', 'step_1', 'step_2', 'step_3', 'confirm']].drop_duplicates()

# Merge the variation column to the pivot_df
pivot_df = pivot_df.merge(step_time_df, on='client_id', how='left')

# Display the updated pivot_df
pivot_df

client_id  start_counts  step_1_counts  step_2_counts  step_3_counts  \
0             555             1              1              1              1   
1             647             1              1              1              1   
2             934             4              0              0              0   
3            1028             1              5              2              1   
4            1104             1              0              0              0   
...           ...           ...            ...            ...            ...   
121981    9999729             1              1              1              1   
121982    9999729             2              1              1              0   
121983    9999729             2              1              1              0   
121984    9999729             2              1              1              0   
121985    9999832             1              1              0              0   

        confirm_counts  visit_count variation  start  step_1  step_2  step_3  \
0                    1            1      Test    7.0    32.0    99.0    20.0   
1                    1            1      Test    7.0    18.0   189.0   163.0   
2                    0            1      Test  142.0     0.0     0.0     0.0   
3                    0            1   Control   49.0    94.0   284.0   111.0   
4                    0            2   Control    0.0     0.0     0.0     0.0   
...                ...          ...       ...    ...     ...     ...     ...   
121981               1            3      Test   15.0     0.0     0.0     0.0   
121982               0            3      Test   82.0   287.0    81.0     0.0   
121983               0            3      Test    5.0    10.0    39.0    21.0   
121984               0            3      Test   15.0     0.0     0.0     0.0   
121985               0            1      Test    8.0     0.0     0.0     0.0   

        confirm  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0  
...         ...  
121981      0.0  
121982      0.0  
121983      0.0  
121984      0.0  
121985      0.0  

[121986 rows x 13 columns]

In [30]:
print(tabla_analisis)

        client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  \
0         3561384             4.0              56.0      59.5     U   
1         3561384             4.0              56.0      59.5     U   
2         7338123             7.0              88.0      23.5     M   
3         7338123             7.0              88.0      23.5     M   
4         7338123             7.0              88.0      23.5     M   
...           ...             ...               ...       ...   ...   
321190    6627522            18.0             227.0      58.0     F   
321191    6627522            18.0             227.0      58.0     F   
321192    6627522            18.0             227.0      58.0     F   
321193    6627522            18.0             227.0      58.0     F   
321194    6627522            18.0             227.0      58.0     F   

        num_accts         bal  calls_6_mnth  logons_6_mnth variation  \
0             2.0    63130.44           6.0            9.0      Test   
1  

In [31]:
# Filtrar las filas donde 'process_step' es igual a 'confirm'
clients_at_confirm = pivot_df[pivot_df['confirm_counts'] > 0]
clients_at_confirm


client_id  start_counts  step_1_counts  step_2_counts  step_3_counts  \
0             555             1              1              1              1   
1             647             1              1              1              1   
10           1195             1              1              1              1   
11           1197             1              1              2              2   
12           1336             1              1              1              1   
...           ...           ...            ...            ...            ...   
121972    9999009             1              1              1              1   
121974    9999400             1              1              1              1   
121979    9999729             1              1              1              1   
121980    9999729             1              1              1              1   
121981    9999729             1              1              1              1   

        confirm_counts  visit_count variation  start  step_1  step_2  step_3  \
0                    1            1      Test    7.0    32.0    99.0    20.0   
1                    1            1      Test    7.0    18.0   189.0   163.0   
10                   1            1   Control   33.0    98.0    31.0    83.0   
11                   1            1   Control    4.0     6.0    57.0    28.0   
12                   1            2      Test   25.0    26.0    51.0   109.0   
...                ...          ...       ...    ...     ...     ...     ...   
121972               1            1      Test    8.0    26.0    19.0   118.0   
121974               1            1      Test   22.0    27.0    46.0    24.0   
121979               1            3      Test   82.0   287.0    81.0     0.0   
121980               1            3      Test    5.0    10.0    39.0    21.0   
121981               1            3      Test   15.0     0.0     0.0     0.0   

        confirm  
0           0.0  
1           0.0  
10          0.0  
11          0.0  
12          0.0  
...         ...  
121972      0.0  
121974      0.0  
121979      0.0  
121980      0.0  
121981      0.0  

[57359 rows x 13 columns]

In [32]:
confirm = len(clients_at_confirm)
start = len(clients_at_start)
step1 = len(clients_at_step_1)
step2 = len(clients_at_step_2)
step3 = len(clients_at_step_3)
total_usuarios = len(df_final_demo)
total_test = len(pivot_df)

NameError: name 'clients_at_start' is not defined

In [ ]:
total_usuarios, total_test, start ,step1, step2 ,step3 , confirm

In [ ]:
d_index = {
    "name": ['total_usuarios', 'total_test', 'start' ,'step1', 'step2' ,'step3' , 'confirm'],
    "score": [total_usuarios, total_test, start ,step1, step2 ,step3 , confirm]
}

df_index = pd.DataFrame(d_index)
df_index

TOTAL CLIENTES= 70,609
START = 50,063
STEP 1 = 44,406
STEP 2 = 40,896
STEP 3 = 38,292
CONFIRM = 34,110
% de caída de START a STEP 1: 11.37%
% de caída de STEP 1 a STEP 2: 7.97%
% de caída de STEP 2 a STEP 3: 6.37%
% de caída de STEP 3 a CONFIRM: 10.92%
% de clientes que han confirmado: 48.3%

In [ ]:
pivot_df.sum()

In [ ]:
tabla_analisis.describe()

In [ ]:
tabla_analisis

In [ ]:
# Step 1: Filter the DataFrame to include only rows where 'time_in_step' is greater than 0 and less than 2 seconds
users_less_than_2_seconds = tabla_analisis[(tabla_analisis['time_in_step'] > 0) & (tabla_analisis['time_in_step'] < 2)]

# Step 2: Count the number of occurrences for each user
user_counts = users_less_than_2_seconds['client_id'].value_counts().reset_index()
user_counts.columns = ['client_id', 'error_count']

# Step 3: Merge this count back into the pivot_df
pivot_df = pivot_df.merge(user_counts, on='client_id', how='left')

# Fill NaN values with 0 (if any user didn't have any step with time_in_step between 0 and 2 seconds)
pivot_df['error_count'] = pivot_df['error_count'].fillna(0)

# Display the updated pivot_df
pivot_df



In [ ]:
# Filter the pivot_df to include only rows where 'error_count' is greater than 0
errors_greater_than_zero = pivot_df[pivot_df['error_count'] > 0]

# Display the filtered DataFrame
errors_greater_than_zero

In [ ]:
# Ensure the DataFrame is sorted by 'visit_id' and 'date_time'
tabla_analisis_sorted = tabla_analisis.sort_values(by=['visit_id', 'date_time'])

# Define the order of process steps
process_order = {'start': 1, 'step_1': 2, 'step_2': 3, 'step_3': 4, 'confirm': 5}

# Map the process steps to their order
tabla_analisis_sorted['process_step_order'] = tabla_analisis_sorted['process_step'].map(process_order)

# Function to count backward steps in the process
def count_backward_steps(group):
    return (group['process_step_order'].diff() < 0).sum()

# Group by 'visit_id' and apply the function
backward_steps_per_visit = tabla_analisis_sorted.groupby('visit_id').apply(count_backward_steps).reset_index()
backward_steps_per_visit.columns = ['visit_id', 'backward_steps']

# Merge backward steps count back to the original DataFrame
tabla_analisis_with_backward = tabla_analisis_sorted.merge(backward_steps_per_visit, on='visit_id', how='left')

# Group by 'client_id' to get the total number of backward steps per user
backward_steps_per_user = tabla_analisis_with_backward.groupby('client_id')['backward_steps'].sum().reset_index()
backward_steps_per_user.columns = ['client_id', 'backward_steps_count']

# Merge this count back into the pivot_df
pivot_df = pivot_df.merge(backward_steps_per_user, on='client_id', how='left')

# Fill NaN values with 0 (if any user didn't have any backward steps)
pivot_df['backward_steps_count'] = pivot_df['backward_steps_count'].fillna(0)

# Display the updated pivot_df
pivot_df

In [ ]:
# Filter the pivot_df to include only rows where 'error_count' is greater than 0
backward_steps_greater_than_zero = pivot_df[pivot_df['backward_steps_count'] > 0]

# Display the filtered DataFrame
backward_steps_greater_than_zero

Answer the following questions about demographics:

Who are the primary clients using this online process?
Are the primary clients younger or older, new or long-standing?

In [ ]:
#Are the primary clients younger or older

# Calculate the mean age of the clients
mean_age = tabla_analisis['clnt_age'].mean()

# Print the mean age with 2 decimal places
print(f"La media de edad es de {mean_age:.2f}")

# Calculate the median tenure of the clients
median_age = tabla_analisis['clnt_age'].median()

# Print the median tenure with 2 decimal places
print(f"La mediana de edad es de {median_age:.2f}")

In [ ]:
#New or long-standing?

# Calculate the mean age of the clients
mean_standing_time = tabla_analisis['clnt_tenure_mnth'].mean()

# Print the mean age with 2 decimal places
print(f"La media de meses de estancia en la plataforma es de {mean_standing_time:.2f}")

# Calculate the median tenure of the clients
median_standing_time = tabla_analisis['clnt_tenure_mnth'].median()

# Print the median tenure with 2 decimal places
print(f"La mediana de meses de estancia en la plataforma es de {median_standing_time:.2f}")


In [ ]:
tabla_analisis.head()

In [ ]:
# Filter the DataFrame to include only rows where 'client_id' is 836976
client_rows = tabla_analisis[tabla_analisis['client_id'] == 6002181]

# Display the filtered rows
client_rows

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  \
65        6002181            17.0             214.0      48.0     M   
70695     6002181            17.0             214.0      48.0     M   
70696     6002181            17.0             214.0      48.0     M   
70697     6002181            17.0             214.0      48.0     M   
70698     6002181            17.0             214.0      48.0     M   
70699     6002181            17.0             214.0      48.0     M   
70700     6002181            17.0             214.0      48.0     M   
70701     6002181            17.0             214.0      48.0     M   
226286    6002181            17.0             214.0      48.0     M   
260577    6002181            17.0             214.0      48.0     M   
260578    6002181            17.0             214.0      48.0     M   
260579    6002181            17.0             214.0      48.0     M   

        num_accts        bal  calls_6_mnth  logons_6_mnth            bal_str  \
65            3.0  746472.21           1.0            4.0  746472.2100000001   
70695         3.0  746472.21           1.0            4.0  746472.2100000001   
70696         3.0  746472.21           1.0            4.0  746472.2100000001   
70697         3.0  746472.21           1.0            4.0  746472.2100000001   
70698         3.0  746472.21           1.0            4.0  746472.2100000001   
70699         3.0  746472.21           1.0            4.0  746472.2100000001   
70700         3.0  746472.21           1.0            4.0  746472.2100000001   
70701         3.0  746472.21           1.0            4.0  746472.2100000001   
226286        3.0  746472.21           1.0            4.0  746472.2100000001   
260577        3.0  746472.21           1.0            4.0  746472.2100000001   
260578        3.0  746472.21           1.0            4.0  746472.2100000001   
260579        3.0  746472.21           1.0            4.0  746472.2100000001   

        ...             visitor_id                      visit_id  \
65      ...   203800014_3409544458  100221419_95702886123_228743   
70695   ...  727235195_74792444362  298896487_11483635593_672780   
70696   ...  727235195_74792444362  298896487_11483635593_672780   
70697   ...  727235195_74792444362  298896487_11483635593_672780   
70698   ...  727235195_74792444362  298896487_11483635593_672780   
70699   ...  727235195_74792444362  298896487_11483635593_672780   
70700   ...  727235195_74792444362  298896487_11483635593_672780   
70701   ...  727235195_74792444362  298896487_11483635593_672780   
226286  ...  727235195_74792444362  733648203_55673234897_567400   
260577  ...  727235195_74792444362  828949317_28441223423_116452   
260578  ...  727235195_74792444362  828949317_28441223423_116452   
260579  ...  727235195_74792444362  828949317_28441223423_116452   

        process_step           date_time time_in_step confirm start step_1  \
65             start 2017-04-09 07:22:29          NaN     0.0   0.0    0.0   
70695          start 2017-04-09 07:28:03         15.0     0.0  37.0  106.0   
70696          start 2017-04-09 07:28:18         15.0     0.0  37.0  106.0   
70697         step_1 2017-04-09 07:28:33        106.0     0.0  37.0  106.0   
70698         step_2 2017-04-09 07:30:19         65.0     0.0  37.0  106.0   
70699         step_3 2017-04-09 07:31:24        434.0     0.0  37.0  106.0   
70700          start 2017-04-09 07:38:38          7.0     0.0  37.0  106.0   
70701         step_1 2017-04-09 07:38:45          NaN     0.0  37.0  106.0   
226286         start 2017-03-31 17:16:31          NaN     0.0   0.0    0.0   
260577        step_1 2017-03-31 17:16:41         72.0     0.0   0.0   72.0   
260578        step_2 2017-03-31 17:17:53         58.0     0.0   0.0   72.0   
260579        step_3 2017-03-31 17:18:51          NaN     0.0   0.0   72.0   

        step_2  step_3  
65         0.0     0.0  
70695     65.0   434.0  
70696     65.0   434.0  
70697     65.0   434.0  
70698     65.0   4

In [ ]:
# Count the number of occurrences of each unique value in the 'gendr' column
gender_counts = tabla_analisis['gendr'].value_counts()

# Print the counts of each gender
print(gender_counts)

In [ ]:
#Completion Rate: The proportion of users who reach the final ‘confirm’ step.

completion_counts = tabla_analisis['process_step'].value_counts()
print(completion_counts)

In [ ]:
tabla_analisis['variation'].value_counts()

In [374]:
# Filter the DataFrame to include only rows where the step is 'confirm'
confirm_clients = tabla_analisis[tabla_analisis['process_step'] == 'confirm']

# Get the unique client IDs who reached the 'confirm' step
unique_confirm_clients = confirm_clients['client_id'].unique()

In [ ]:
len(unique_confirm_clients)

In [ ]:
unique_clients_at_confirm=(len(unique_confirm_clients)/len(tabla_analisis['client_id'].unique()))*100
print(f"La tasa de finalización es de {unique_clients_at_confirm:.2f}")

In [ ]:
confirm_clients_test = confirm_clients[confirm_clients['variation'] == 'Test']
confirm_clients_control = confirm_clients[confirm_clients['variation'] == 'Control']
len(confirm_clients_test), len(confirm_clients_control)

In [ ]:
# Get the unique client IDs who reached the 'confirm' step from each group
unique_confirm_clients_test = confirm_clients_test['client_id'].unique()
unique_confirm_clients_control = confirm_clients_control['client_id'].unique()
len(unique_confirm_clients_test), len(unique_confirm_clients_control)

Based on the chosen KPIs, how does the new design’s performance compare to the old one?

The new design (test) has more number confirms than the older (control)

(25716 VS 17498)

And the % is higher

variation
Test       177787
Control    143408
Name: count, dtype: int64

14.46% Test
12.20% Control

In [ ]:
tabla_analisis.dtypes

In [ ]:
# Convert the timestamp column to datetime
tabla_analisis['date_time'] = pd.to_datetime(tabla_analisis['date_time'])

# Sort the DataFrame by user_id and timestamp
user_steps = tabla_analisis.sort_values(by=['client_id', 'date_time'])

# Calculate the time spent on each step by each user
user_steps['date_time'] = user_steps.groupby('client_id')['date_time'].diff().shift(-1)

# Calculate the average time spent on each step
average_time_spent = user_steps.groupby('process_step')['date_time'].mean()

# Print the average time spent on each step
print("Average Time Spent on Each Step:")
print(average_time_spent)

In [ ]:
# Calculate the average time spent on each step
median_time_spent = user_steps.groupby('process_step')['date_time'].median()

# Print the average time spent on each step
print("Median Value of Time Spent on Each Step:")
print(median_time_spent)

In [ ]:
# Sort the DataFrame by user_id and timestamp
user_steps_visit_id = tabla_analisis.sort_values(by=['client_id', 'visit_id', 'date_time'])

# Calculate the time spent on each step by each user
user_steps_visit_id['date_time'] = user_steps_visit_id.groupby('visit_id')['date_time'].diff().shift(-1)

# Calculate the average time spent on each step
average_time_spent = user_steps_visit_id.groupby('process_step')['date_time'].mean()

# Print the average time spent on each step
print("Average Time Spent on Each Step:")
print(average_time_spent)

In [ ]:
print(user_steps_visit_id.head(100))

In [ ]:
# Display the first 100 rows of the user_steps DataFrame
user_steps.head(10)

In [ ]:
tabla_analisis.head()

In [ ]:
# Filter the DataFrame to include only rows where 'client_id' is 836976
client_rows = user_steps[user_steps['client_id'] == 1516]

# Display the filtered rows
client_rows

In [387]:
test_group = tabla_analisis[tabla_analisis['variation'] == 'Test']
control_group = tabla_analisis[tabla_analisis['variation'] == 'Control']

In [ ]:
test_group, control_group

In [ ]:
# Sort the DataFrame by 'client_id', 'visit_id', and 'date_time'
user_steps_test = test_group.sort_values(by=['client_id', 'visit_id', 'date_time'])

# Calculate the time spent on each step by each user within each visit
user_steps_test['time_spent'] = user_steps_test.groupby(['client_id', 'visit_id'])['date_time'].diff().shift(-1)

# Calculate the average time spent on each step
average_time_spent = user_steps_test.groupby('process_step')['time_spent'].mean()

# Print the average time spent on each step
print("Average Time Spent on Each Step for Test group:")
print(average_time_spent)

In [ ]:
# Sort the DataFrame by user_id and timestamp
user_steps_control = control_group.sort_values(by=['client_id', 'date_time'])

# Calculate the time spent on each step by each user
user_steps_control['date_time'] = user_steps_control.groupby(['client_id', 'visit_id'])['date_time'].diff().shift(-1)

# Calculate the average time spent on each step
average_time_spent = user_steps_control.groupby('process_step')['date_time'].mean()

# Print the average time spent on each step
print("Average Time Spent on Each Step for Control group:")
print(average_time_spent)

In [ ]:
# Filter the DataFrame to include only rows where time spent is greater than 10 hours
ten_hours = pd.Timedelta(hours=10)
filtered_users = user_steps[user_steps['date_time'] > ten_hours]

# Display the filtered rows
len(filtered_users['client_id'].unique())

In [ ]:
# Filter the DataFrame to include only rows where time spent is greater than 10 hours
one_hour = pd.Timedelta(hours=1)
filtered_users = user_steps[user_steps['date_time'] < one_hour]

# Display the filtered rows
len(filtered_users['client_id'].unique())

In [ ]:
tabla_analisis.head()

In [ ]:
# Sort the DataFrame by 'client_id', 'visit_id', and 'date_time'
df_sorted = tabla_analisis.sort_values(by=['client_id', 'visit_id', 'date_time'])

# Define the order of steps
step_order = {'start': 1, 'step_1': 2, 'step_2': 3, 'step_3': 4, 'confirm': 5}

# Map the process steps to their order
df_sorted['step_order'] = df_sorted['process_step'].map(step_order)

# Identify instances where users move from a later step to an earlier step
df_sorted['previous_step_order'] = df_sorted.groupby(['client_id', 'visit_id'])['step_order'].shift(1)
df_sorted['error'] = df_sorted['step_order'] < df_sorted['previous_step_order']

# Count the number of errors
error_count = df_sorted['error'].sum()

# Calculate the total number of transitions
total_transitions = df_sorted['error'].count() - df_sorted.groupby(['client_id', 'visit_id']).ngroups

# Calculate the error rate
error_rate = error_count / total_transitions

# Display the error rate
print(f"Error Rate: {error_rate:.2%}")

# Display the DataFrame with error identification
df_sorted[['client_id', 'visit_id', 'process_step', 'date_time', 'step_order', 'previous_step_order', 'error']]

In [ ]:
# Filter the DataFrame to show only rows where 'error' is True
errors_df = df_sorted[df_sorted['error'] == True]

# Display the filtered DataFrame
errors_df[['client_id', 'visit_id', 'process_step', 'date_time', 'step_order', 'previous_step_order', 'error']]

In [ ]:
# Sort the DataFrame by 'client_id', 'visit_id', and 'date_time'
df_sorted_test = test_group.sort_values(by=['client_id', 'visit_id', 'date_time'])

# Define the order of steps
step_order = {'start': 1, 'step_1': 2, 'step_2': 3, 'step_3': 4, 'confirm': 5}

# Map the process steps to their order
df_sorted_test['step_order'] = df_sorted_test['process_step'].map(step_order)

# Identify instances where users move from a later step to an earlier step
df_sorted_test['previous_step_order'] = df_sorted_test.groupby(['client_id', 'visit_id'])['step_order'].shift(1)
df_sorted_test['error'] = df_sorted_test['step_order'] < df_sorted_test['previous_step_order']

# Count the number of errors
error_count = df_sorted_test['error'].sum()

# Calculate the total number of transitions
total_transitions = df_sorted_test['error'].count() - df_sorted_test.groupby(['client_id', 'visit_id']).ngroups

# Calculate the error rate
error_rate = error_count / total_transitions

# Display the error rate
print(f"Error Rate: {error_rate:.2%}")

# Display the DataFrame with error identification
df_sorted_test[['client_id', 'visit_id', 'process_step', 'date_time', 'step_order', 'previous_step_order', 'error']]

In [ ]:

# Sort the DataFrame by 'client_id', 'visit_id', and 'date_time'
df_sorted_control = control_group.sort_values(by=['client_id', 'visit_id', 'date_time'])

# Define the order of steps
step_order = {'start': 1, 'step_1': 2, 'step_2': 3, 'step_3': 4, 'confirm': 5}

# Map the process steps to their order
df_sorted_control['step_order'] = df_sorted_control['process_step'].map(step_order)

# Identify instances where users move from a later step to an earlier step
df_sorted_control['previous_step_order'] = df_sorted_control.groupby(['client_id', 'visit_id'])['step_order'].shift(1)
df_sorted_control['error'] = df_sorted_control['step_order'] < df_sorted_control['previous_step_order']

# Count the number of errors
error_count = df_sorted_control['error'].sum()

# Calculate the total number of transitions
total_transitions = df_sorted_control['error'].count() - df_sorted_control.groupby(['client_id', 'visit_id']).ngroups

# Calculate the error rate
error_rate = error_count / total_transitions

# Display the error rate
print(f"Error Rate: {error_rate:.2%}")

# Display the DataFrame with error identification
df_sorted_control[['client_id', 'visit_id', 'process_step', 'date_time', 'step_order', 'previous_step_order', 'error']]

H4.

The new design (test) has more number confirms than the older (control)

(Test 25716 VS Control 17498)

And the % is higher

variation
Test       177787
Control    143408
Name: count, dtype: int64

14.46% Test
12.20% Control

To conduct hypothesis testing on the completion rate between the Test and Control groups, follow these steps:

Define the Hypotheses:

Null Hypothesis (H0): There is no difference in the completion rates between the Test and Control groups.
Alternative Hypothesis (H1): There is a difference in the completion rates between the Test and Control groups.
Choose the Significance Level:

Commonly used significance level (α) is 0.05.
Calculate the Completion Rates:

Calculate the completion rates for both the Test and Control groups.
Perform the Statistical Test:

Use a two-proportion z-test to compare the completion rates between the two groups.
Interpret the Results:

Determine if you can reject the null hypothesis based on the p-value and the significance level.

In [398]:
import scipy.stats as st
from scipy.stats import ttest_ind

In [ ]:
tabla_analisis

In [ ]:
test_group

In [401]:
test_confirmed = test_group['process_step'].apply(lambda x: 1 if x == 'confirm' else 0)
control_confirmed = control_group['process_step'].apply(lambda x: 1 if x == 'confirm' else 0)

In [ ]:
test_confirmed, control_confirmed

In [ ]:
# Perform the Two-sample t-test
stat, p_value = ttest_ind(test_confirmed, control_confirmed, equal_var=False)

# Define the significance level
alpha = 0.05

# Print the results
print(f"Test Statistic: {stat}")
print(f"P-value: {p_value}")

if p_value < alpha:
    print("Reject the null hypothesis: There is a significant difference in completion rates between the Test and Control groups.")
else:
    print("Fail to reject the null hypothesis: There is no significant difference in completion rates between the Test and Control groups.")

### Conclusion:

Based on the results of the Two-sample t-test:

- **Test Statistic**: 18.83876734309717
- **P-value**: 4.0175581234786464e-79

Since the p-value is significantly less than the commonly used significance level of 0.05, we reject the null hypothesis.

### Interpretation:
- **Reject the Null Hypothesis**: There is a statistically significant difference in completion rates between the Test and Control groups.
- **Significance**: The extremely low p-value indicates that the observed difference in completion rates is highly unlikely to have occurred by chance.

### Practical Implications:
- **Effectiveness of the Redesign**: The new design (Test group) has a significantly higher completion rate compared to the old design (Control group).
- **Decision Making**: Based on this analysis, it is recommended to adopt the new design, as it leads to a higher completion rate, which can positively impact user engagement and overall performance metrics.

### Summary:
The hypothesis testing confirms that the new design is more effective in terms of completion rates. The significant difference observed suggests that the redesign has successfully improved user experience and completion rates. Therefore, the new design should be implemented to achieve better outcomes.

H5.

Completion Rate with a Cost-Effectiveness Threshold
The introduction of a new UI design comes with its associated costs: design, development, testing, potential training for staff, and possible short-term disruptions or adjustments for users. To justify these costs, Vanguard has determined that any new design should lead to a minimum increase in the completion rate to be deemed cost-effective.

Threshold: Vanguard has set this minimum increase in completion rate at 5%. This is the rate at which the projected benefits, in terms of increased user engagement and potential revenue, are estimated to outweigh the costs of the new design.

You are required to carry out another analysis, ensuring that the observed increase in completion rate from the A/B test meets or exceeds this 5% threshold. If the new design doesn’t lead to at least this level of improvement, it may not be justifiable from a cost perspective, regardless of its statistical significance.

In [ ]:
# Given data
test_completions = 25716
control_completions = 17498
test_total = 177787
control_total = 143408

# Calculate completion rates
completion_rate_test = test_completions / test_total
completion_rate_control = control_completions / control_total

# Calculate the observed increase in completion rate
observed_increase = completion_rate_test - completion_rate_control

# Define the threshold
threshold = 0.05

# Perform the one-sided two-proportion z-test
count = [test_completions, control_completions]
nobs = [test_total, control_total]
stat, p_value = proportions_ztest(count, nobs, alternative='larger')

# Define the significance level
alpha = 0.05

# Print the results
print(f"Completion Rate (Test): {completion_rate_test:.4f}")
print(f"Completion Rate (Control): {completion_rate_control:.4f}")
print(f"Observed Increase: {observed_increase:.4f}")
print(f"Test Statistic: {stat}")

### Conclusions:

Based on the results of the one-sided two-proportion z-test:

- **Completion Rate (Test)**: 14.46%
- **Completion Rate (Control)**: 12.20%
- **Observed Increase**: 2.26%
- **Test Statistic**: -18.684265189201895
- **P-value**: 1.0

### Interpretation:
- **Fail to Reject the Null Hypothesis**: The observed increase in completion rate does not meet the 5% threshold and is not statistically significant.
- **Threshold Not Met**: The observed increase of 2.26% is below the required 5% threshold set by Vanguard to justify the costs associated with the new design.

### Practical Implications:
- **Cost-Effectiveness**: The new design does not lead to a sufficient increase in completion rate to be deemed cost-effective. The increase in completion rate is not large enough to outweigh the costs of the new design.
- **Decision Making**: Based on this analysis, it is not recommended to adopt the new design solely based on the completion rate improvement, as it does not meet the cost-effectiveness threshold.

### Summary:
The hypothesis testing confirms that the new design does not achieve the required 5% increase in completion rate to justify its costs. Although the new design has a higher completion rate compared to the old design, the increase is not statistically significant and does not meet the threshold set by Vanguard. Therefore, the new design may not be justifiable from a cost perspective.

H6.

You have been given the freedom to choose another hypothesis to test. Here are some examples:

-You might want to test whether the average age of clients engaging with the new process is the same as those engaging with the old process

-You might want to test if the average client tenure (how long they’ve been with Vanguard) of those engaging with the new process is the same as those engaging with the old process

-You might want to test if there are gender differences that affect engaging with the new or old process

Make sure to define the proper null and alternative hypothesis to test it. You are required to choose one or come up with another of your own to test.

In [ ]:
test_group

In [ ]:
# Assuming test_group and control_group DataFrames are already defined
# Sample data structure
# test_group = pd.DataFrame({'client_id': [...], 'clnt_age': [...], 'process_step': [...]})
# control_group = pd.DataFrame({'client_id': [...], 'clnt_age': [...], 'process_step': [...]})

# Filter the data to get the client age for each group
test_age = test_group['clnt_age']
control_age = control_group['clnt_age']

# Perform the Two-sample t-test
stat, p_value = ttest_ind(test_age, control_age, equal_var=False)

# Define the significance level
alpha = 0.05

# Print the results
print(f"Test Statistic: {stat}")
print(f"P-value: {p_value}")

if p_value < alpha:
    print("Reject the null hypothesis: There is a significant difference in average age between the Test and Control groups.")
else:
    print("Fail to reject the null hypothesis: There is no significant difference in average age between the Test and Control groups.")

In [ ]:
# Assuming test_group and control_group DataFrames are already defined
# Sample data structure
# test_group = pd.DataFrame({'client_id': [...], 'clnt_tenure_mnth': [...], 'process_step': [...]})
# control_group = pd.DataFrame({'client_id': [...], 'clnt_tenure_mnth': [...], 'process_step': [...]})

# Filter the data to get the client tenure for each group
test_tenure = test_group['clnt_tenure_mnth']
control_tenure = control_group['clnt_tenure_mnth']

# Perform the Two-sample t-test
stat, p_value = ttest_ind(test_tenure, control_tenure, equal_var=False)

# Define the significance level
alpha = 0.05

# Print the results
print(f"Test Statistic: {stat}")
print(f"P-value: {p_value}")

if p_value < alpha:
    print("Reject the null hypothesis: There is a significant difference in average client tenure between the Test and Control groups.")
else:
    print("Fail to reject the null hypothesis: There is no significant difference in average client tenure between the Test and Control groups.")

In [ ]:
from scipy.stats import chi2_contingency

# Assuming test_group and control_group DataFrames are already defined
# Sample data structure
# test_group = pd.DataFrame({'client_id': [...], 'gendr': [...], 'process_step': [...]})
# control_group = pd.DataFrame({'client_id': [...], 'gendr': [...], 'process_step': [...]})

# Create a contingency table for gender and group
contingency_table = pd.crosstab(tabla_analisis['variation'], tabla_analisis['gendr'])

# Perform the Chi-square test
chi2, p_value, dof, expected = chi2_contingency(contingency_table)

# Define the significance level
alpha = 0.05

# Print the results
print(f"Chi-square Statistic: {chi2}")
print(f"P-value: {p_value}")

if p_value < alpha:
    print("Reject the null hypothesis: There is a significant difference in gender proportions between the Test and Control groups.")
else:
    print("Fail to reject the null hypothesis: There is no significant difference in gender proportions between the Test and Control groups.")

### Conclusions:

Based on the results of the hypothesis tests, we can draw the following conclusions:

#### 1. Average Age of Clients:
- **Test Statistic**: 7.928355666370457
- **P-value**: 2.228041702345781e-15
- **Conclusion**: Reject the null hypothesis.

**Interpretation**:
- There is a statistically significant difference in the average age of clients between the Test and Control groups.
- This suggests that the new process (Test group) attracts clients of a different age profile compared to the old process (Control group).

#### 2. Average Client Tenure:
- **Test Statistic**: 0.505067363760497
- **P-value**: 0.6135118864992013
- **Conclusion**: Fail to reject the null hypothesis.

**Interpretation**:
- There is no statistically significant difference in the average client tenure between the Test and Control groups.
- This suggests that the new process (Test group) does not significantly differ from the old process (Control group) in terms of how long clients have been with Vanguard.

#### 3. Gender Differences:
- **Chi-square Statistic**: 54.73241150360589
- **P-value**: 7.830657522268704e-12
- **Conclusion**: Reject the null hypothesis.

**Interpretation**:
- There is a statistically significant difference in gender proportions between the Test and Control groups.
- This suggests that the new process (Test group) engages a different gender distribution compared to the old process (Control group).

### Practical Implications:
1. **Average Age**:
   - The significant difference in average age indicates that the new design may appeal more to a specific age group. This insight can be used to tailor marketing strategies and user experience improvements to better serve this demographic.

2. **Average Client Tenure**:
   - Since there is no significant difference in client tenure, the new design does not appear to affect how long clients have been with Vanguard. This suggests that the new design is equally engaging for both new and long-term clients.

3. **Gender Differences**:
   - The significant difference in gender proportions suggests that the new design may be more appealing to a particular gender. This insight can be used to further refine the design and marketing efforts to ensure inclusivity and appeal to a broader audience.

### Summary:
The hypothesis testing reveals significant differences in average age and gender proportions between the Test and Control groups, while there is no significant difference in average client tenure. These insights can guide further improvements and targeted strategies to enhance the effectiveness of the new design.

### Experiment Evaluation

#### Design Effectiveness

1. **Was the experiment well-structured?**
   - **Answer**: The experiment appears to be well-structured as it involves a clear comparison between the Test group (new design) and the Control group (old design). The use of statistical tests to compare completion rates, average age, client tenure, and gender proportions indicates a thorough approach to evaluating the effectiveness of the new design.

2. **Were clients randomly and equally divided between the old and new designs?**
   - **Answer**: To determine if clients were randomly and equally divided, we would need to verify the randomization process used in the experiment. If the assignment to Test and Control groups was done randomly and the sample sizes are large enough, we can assume that the division was fair. The provided data shows a substantial number of clients in both groups (177,787 in Test and 143,408 in Control), which suggests a reasonably balanced experiment.

3. **Were there any biases?**
   - **Answer**: Potential biases could include selection bias if the clients were not randomly assigned to the Test and Control groups. Additionally, if certain demographics (e.g., age, gender) were overrepresented in one group, it could introduce bias. The significant differences in average age and gender proportions between the groups suggest that there might have been some underlying biases in the assignment process. Ensuring proper randomization and stratification can help mitigate these biases.

#### Duration Assessment

1. **Was the timeframe of the experiment (from 3/15/2017 to 6/20/2017) adequate to gather meaningful data and insights?**
   - **Answer**: The experiment duration of approximately three months (from 3/15/2017 to 6/20/2017) seems adequate to gather meaningful data and insights. This timeframe allows for the collection of a substantial amount of data, as evidenced by the large sample sizes in both the Test and Control groups. It also provides enough time to observe user behavior and engagement with the new design.

#### Additional Data Needs

1. **What other data, if available, could enhance the analysis?**
   - **Answer**: Additional data that could enhance the analysis includes:
     - **Demographic Information**: More detailed demographic data (e.g., income level, education) could help identify specific segments of the population that respond differently to the new design.
     - **Behavioral Data**: Information on user interactions with the website (e.g., click-through rates, time spent on each page) could provide deeper insights into how the new design affects user behavior.
     - **Feedback and Satisfaction Scores**: Collecting user feedback and satisfaction scores could help understand the qualitative aspects of the user experience with the new design.
     - **Conversion Funnel Data**: Data on different stages of the conversion funnel (e.g., from initial visit to final conversion) could help identify where users drop off and how the new design impacts each stage.
     - **A/B Test Metadata**: Information on how the randomization was performed and any stratification criteria used could help assess the fairness and validity of the experiment.

### Summary

The experiment appears to be well-structured with a clear comparison between the Test and Control groups. The duration of the experiment seems adequate to gather meaningful data. However, potential biases in the assignment process should be addressed to ensure the validity of the results. Additional data on demographics, user behavior, feedback, and conversion funnel stages could further enhance the analysis and provide more comprehensive insights into the effectiveness of the new design.

In [32]:
tabla_analisis.to_csv('tabla_analisis.csv', index=False)

In [410]:
test_group.to_csv('test_group.csv', index=False)

In [411]:
control_group.to_csv('control_group.csv', index=False)

In [420]:
pivot_df.to_csv('pivot_df.csv', index=False)

In [ ]:
pivot_df

In [36]:
# Step 1: Ensure the 'date_time' column is in datetime format
tabla_analisis['date_time'] = pd.to_datetime(tabla_analisis['date_time'])

# Step 2: Extract the day of the week from the 'date_time' column
# The day_name() method returns the name of the day (e.g., 'Monday', 'Tuesday')
tabla_analisis['day_of_week'] = tabla_analisis['date_time'].dt.day_name()

# Step 3: Filter the DataFrame to include only rows where 'process_step' is 'confirm'
confirm_steps = tabla_analisis[tabla_analisis['process_step'] == 'confirm']

# Step 4: Count the occurrences of each day of the week for the 'confirm' steps
confirm_day_counts = confirm_steps['day_of_week'].value_counts()

# Display the sorted list
confirm_day_counts

day_of_week
Wednesday    14381
Thursday      6173
Sunday        4823
Monday        4815
Friday        4621
Tuesday       4480
Saturday      3921
Name: count, dtype: int64